compiler/qsc_linter/src/linter.rs

```diff
@@ -0,0 +1,67 @@
+pub(crate) mod ast;
+pub(crate) mod hir;
+
+use miette::{Diagnostic, LabeledSpan};
+use qsc_data_structures::span::Span;
+use std::fmt::Display;
+
+/// A lint emited by the linter.
+#[derive(Debug, Clone, thiserror::Error)]
+pub struct Lint {
+    /// A span indicating where the diagnostic is in the source code.
+    pub span: Span,
+    /// This is the message the user will see in the code editor.
+    pub message: &'static str,
+    /// The lint level: allow, warning, error.
+    pub level: LintLevel,
+}
+
+impl std::fmt::Display for Lint {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+
+impl Diagnostic for Lint {
+    fn severity(&self) -> Option<miette::Severity> {
+        match self.level {
+            LintLevel::Allow => None,
+            LintLevel::Warning | LintLevel::ForceWarning => Some(miette::Severity::Warning),
+            LintLevel::Error | LintLevel::ForceError => Some(miette::Severity::Error),
+        }
+    }
+
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
+        let source_span = miette::SourceSpan::from(self.span);
+        let labeled_span = LabeledSpan::new_with_span(Some(self.to_string()), source_span);
+        Some(Box::new(vec![labeled_span].into_iter()))
+    }
+}
+
+/// A lint level. This defines if a lint will be treated as a warning or an error,
+/// and if the lint level can be overriden by the user.
+#[derive(Debug, Clone)]
+pub enum LintLevel {
```

Comment:
Is ForceWarning and ForceError common among other linters? If there is not a precedent for this, it doesn't seem like something that should be used, as users should be able to control how to treat the lints.

Reply:
Yes, for example, it is used in the rust linter, for example, to force future deprecation lints, some stylistic conventions, and things like dead-code.